<a href="https://colab.research.google.com/github/Jack-the-first/LaTeX-Workshop/blob/master/Task_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Load Data 

In [3]:
import pandas as pd
adult_primary=pd.read_csv('adult.data')
banknote_primary=pd.read_csv('data_banknote_authentication.txt')
titanic_primary=pd.read_csv('train_and_test2.csv')

In [4]:
adult_primary.head()

39          State-gov   77516   Bachelors   13        Never-married  \
0  50   Self-emp-not-inc   83311   Bachelors   13   Married-civ-spouse   
1  38            Private  215646     HS-grad    9             Divorced   
2  53            Private  234721        11th    7   Married-civ-spouse   
3  28            Private  338409   Bachelors   13   Married-civ-spouse   
4  37            Private  284582     Masters   14   Married-civ-spouse   

         Adm-clerical   Not-in-family   White     Male   2174   0   40  \
0     Exec-managerial         Husband   White     Male      0   0   13   
1   Handlers-cleaners   Not-in-family   White     Male      0   0   40   
2   Handlers-cleaners         Husband   Black     Male      0   0   40   
3      Prof-specialty            Wife   Black   Female      0   0   40   
4     Exec-managerial            Wife   White   Female      0   0   40   

    United-States   <=50K  
0   United-States   <=50K  
1   United-States   <=50K  
2   United-States   <=50K  
3            Cuba   <=50K  
4   United-States   <=50K

In [5]:
banknote_primary.head()

3.6216  8.6661  -2.8073  -0.44699  0
0  4.54590  8.1674  -2.4586  -1.46210  0
1  3.86600 -2.6383   1.9242   0.10645  0
2  3.45660  9.5228  -4.0112  -3.59440  0
3  0.32924 -4.4552   4.5718  -0.98880  0
4  4.36840  9.6718  -3.9606  -3.16250  0

In [6]:
titanic_primary.head()

Passengerid   Age     Fare  Sex  sibsp  zero  zero.1  zero.2  zero.3  \
0            1  22.0   7.2500    0      1     0       0       0       0   
1            2  38.0  71.2833    1      1     0       0       0       0   
2            3  26.0   7.9250    1      0     0       0       0       0   
3            4  35.0  53.1000    1      1     0       0       0       0   
4            5  35.0   8.0500    0      0     0       0       0       0   

   zero.4  ...  zero.12  zero.13  zero.14  Pclass  zero.15  zero.16  Embarked  \
0       0  ...        0        0        0       3        0        0       2.0   
1       0  ...        0        0        0       1        0        0       0.0   
2       0  ...        0        0        0       3        0        0       2.0   
3       0  ...        0        0        0       1        0        0       2.0   
4       0  ...        0        0        0       3        0        0       2.0   

   zero.17  zero.18  2urvived  
0        0        0         0  
1        0        0         1  
2        0        0         1  
3        0        0         1  
4        0        0         0  

[5 rows x 28 columns]

# Load Package and some functions 

## SDV

In [7]:
pip install sdv

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 102 kB 5.1 MB/s 
     |████████████████████████████████| 45 kB 4.8 MB/s 
     |████████████████████████████████| 46 kB 5.5 MB/s 
     |████████████████████████████████| 1.2 MB 61.2 MB/s 
     |████████████████████████████████| 98 kB 9.8 MB/s 
     |████████████████████████████████| 53 kB 2.1 MB/s 
     |████████████████████████████████| 11.2 MB 40.9 MB/s 
     |████████████████████████████████| 38.1 MB 384 kB/s 
     |████████████████████████████████| 944 kB 61.4 MB/s 
     |████████████████████████████████| 281 kB 60.3 MB/s 
     |████████████████████████████████| 636 kB 63.1 MB/s 
     |████████████████████████████████| 2.5 MB 46.7 MB/s 
  Attempting uninstall: scipy
    Found existing installation: scipy 1.4.1
    Uninstalling scipy-1.4.1:
      Successfully uninstalled scipy-1.4.1
  Attempting uninstall: pyyaml
    Found existing installation: P

In [8]:
from sdv.tabular import GaussianCopula
from sdv.tabular import CTGAN
from sdv.tabular import CopulaGAN
from sdv.tabular import TVAE

## DataSynthesizer

In [9]:
pip install DataSynthesizer

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [10]:
from DataSynthesizer.DataDescriber import DataDescriber
from DataSynthesizer.DataGenerator import DataGenerator
from DataSynthesizer.ModelInspector import ModelInspector
from DataSynthesizer.lib.utils import read_json_file, display_bayesian_network

# Adult data

## SDV

### GaussianCopula Model

In [9]:
adult_1= GaussianCopula()
adult_1.fit(adult_primary)
adult_sdv_1=adult_1.sample(200)
adult_sdv_1.head()

/usr/local/lib/python3.7/dist-packages/copulas/univariate/truncated_gaussian.py:45: RuntimeWarning: divide by zero encountered in double_scalars
  a = (self.min - loc) / scale
/usr/local/lib/python3.7/dist-packages/copulas/univariate/truncated_gaussian.py:46: RuntimeWarning: divide by zero encountered in double_scalars
  b = (self.max - loc) / scale


39   State-gov   77516      Bachelors   13        Never-married  \
0  32   Local-gov  182431      Assoc-voc   14   Married-civ-spouse   
1  35     Private  198044   Some-college   12        Never-married   
2  41     Private  128146        HS-grad   11   Married-civ-spouse   
3  33     Private  459081      Bachelors   11        Never-married   
4  68           ?  149905   Some-college   10   Married-civ-spouse   

       Adm-clerical  Not-in-family   White     Male   2174     0   40  \
0             Sales        Husband   White   Female   4127   248   59   
1   Exec-managerial      Own-child   Black   Female  13326   245   23   
2   Exec-managerial      Own-child   White   Female   1723  1203   38   
3   Exec-managerial      Own-child   White   Female   7249   268   45   
4             Sales        Husband   Black   Female    390   719   51   

    United-States   <=50K  
0   United-States   <=50K  
1   United-States    >50K  
2     Philippines   <=50K  
3   United-States   <=50K  
4   United-States    >50K

### CTGAN Model

In [10]:
adult_2= CTGAN()
adult_2.fit(adult_primary)
adult_sdv_2=adult_2.sample(200)
adult_sdv_2.head()

/usr/local/lib/python3.7/dist-packages/sklearn/mixture/_base.py:282: ConvergenceWarning: Initialization 1 did not converge. Try different init parameters, or increase max_iter, tol or check for degenerate data.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/mixture/_base.py:282: ConvergenceWarning: Initialization 1 did not converge. Try different init parameters, or increase max_iter, tol or check for degenerate data.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/mixture/_base.py:282: ConvergenceWarning: Initialization 1 did not converge. Try different init parameters, or increase max_iter, tol or check for degenerate data.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/mixture/_base.py:282: ConvergenceWarning: Initialization 1 did not converge. Try different init parameters, or increase max_iter, tol or check for degenerate data.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/mixture/_base.py:282:

39  State-gov   77516      Bachelors   13        Never-married  \
0  41    Private  270528   Some-college   10             Divorced   
1  41    Private  289783      Bachelors   13   Married-civ-spouse   
2  33    Private  352252        HS-grad    9        Never-married   
3  29    Private  213610        HS-grad    9   Married-civ-spouse   
4  19    Private  139053        HS-grad    9        Never-married   

        Adm-clerical   Not-in-family   White     Male   2174     0   40  \
0     Prof-specialty       Own-child   White   Female      2  1913   23   
1    Exec-managerial         Husband   White     Male      0  1872   49   
2    Farming-fishing   Not-in-family   White   Female     14  1835   49   
3   Transport-moving         Husband   White     Male      2     0   49   
4       Craft-repair       Own-child   Black     Male     10     0   40   

    United-States   <=50K  
0   United-States    >50K  
1   United-States    >50K  
2   United-States   <=50K  
3   United-States   <=50K  
4   United-States   <=50K

### CopulaGAN Model

In [ ]:
adult_3=CopulaGAN()
adult_3.fit(adult_primary)
adult_sdv_3=adult_3.sample(200)
adult_sdv_3.head()

/usr/local/lib/python3.7/dist-packages/sklearn/mixture/_base.py:282: ConvergenceWarning: Initialization 1 did not converge. Try different init parameters, or increase max_iter, tol or check for degenerate data.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/mixture/_base.py:282: ConvergenceWarning: Initialization 1 did not converge. Try different init parameters, or increase max_iter, tol or check for degenerate data.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/mixture/_base.py:282: ConvergenceWarning: Initialization 1 did not converge. Try different init parameters, or increase max_iter, tol or check for degenerate data.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/mixture/_base.py:282: ConvergenceWarning: Initialization 1 did not converge. Try different init parameters, or increase max_iter, tol or check for degenerate data.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/mixture/_base.py:282:

### TVAE Model

In [11]:
adult_4=TVAE()
adult_4.fit(adult_primary)
adult_sdv_4=adult_4.sample(200)
adult_sdv_4.head()

/usr/local/lib/python3.7/dist-packages/sklearn/mixture/_base.py:282: ConvergenceWarning: Initialization 1 did not converge. Try different init parameters, or increase max_iter, tol or check for degenerate data.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/mixture/_base.py:282: ConvergenceWarning: Initialization 1 did not converge. Try different init parameters, or increase max_iter, tol or check for degenerate data.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/mixture/_base.py:282: ConvergenceWarning: Initialization 1 did not converge. Try different init parameters, or increase max_iter, tol or check for degenerate data.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/sklearn/mixture/_base.py:282: ConvergenceWarning: Initialization 1 did not converge. Try different init parameters, or increase max_iter, tol or check for degenerate data.
  ConvergenceWarning,
/usr/local/lib/python3.7/dist-packages/ctgan/data_transformer.py:111

39          State-gov   77516      Bachelors   13           Never-married  \
0  22            Private  184530      Bachelors   13   Married-spouse-absent   
1  31            Private  141335        HS-grad    9   Married-spouse-absent   
2  33   Self-emp-not-inc  114444        HS-grad    9      Married-civ-spouse   
3  59            Private  144975        HS-grad    9                Divorced   
4  65   Self-emp-not-inc   95746   Some-college   10      Married-civ-spouse   

       Adm-clerical   Not-in-family                White     Male   2174   0  \
0      Adm-clerical       Own-child                White   Female      0   0   
1      Craft-repair   Not-in-family   Asian-Pac-Islander     Male      0   0   
2   Exec-managerial            Wife                White     Male      1   0   
3      Adm-clerical       Unmarried                White   Female      3   1   
4                 ?         Husband                White     Male      1   0   

    40   United-States   <=50K  
0   20   United-States   <=50K  
1   40   United-States   <=50K  
2   48   United-States   <=50K  
3   10   United-States   <=50K  
4   12   United-States   <=50K

## DataSynthesizer

In [12]:
input_adult = 'adult.data'

# An attribute is categorical if its domain size is less than this threshold.
threshold_value_adult= 50

### random mode

In [13]:
describer_adult_1= DataDescriber(category_threshold=threshold_value_adult)
describer_adult_1.describe_dataset_in_random_mode(input_adult)
describer_adult_1.save_dataset_description_to_file('des_adult_1')

In [14]:
generator_adult_1 = DataGenerator()
generator_adult_1.generate_dataset_in_random_mode(200, 'des_adult_1')
generator_adult_1.save_synthetic_data('gen_adult_1')
pd.read_csv('gen_adult_1').head()

39     State-gov  77516    Bachelors  13       Never-married  \
0  44  Self-emp-inc      3      1st-4th   9             Widowed   
1  47             ?     42      Masters  11             Widowed   
2  64   Without-pay     65  Prof-school   6   Married-AF-spouse   
3  67  Never-worked     20      7th-8th  10             Widowed   
4  67  Never-worked     36    Preschool  12  Married-civ-spouse   

       Adm-clerical   Not-in-family               White    Male  2174   0  40  \
0      Tech-support  Other-relative               White  Female    10  57  29   
1     Other-service       Own-child  Asian-Pac-Islander  Female     2  68  63   
2    Prof-specialty         Husband               Other    Male    75  45  61   
3   Exec-managerial         Husband  Amer-Indian-Eskimo    Male    46  39  91   
4  Transport-moving   Not-in-family  Amer-Indian-Eskimo    Male    51  24  83   

  United-States  <=50K  
0      Scotland   >50K  
1     Nicaragua   >50K  
2             ?   >50K  
3     Guatemala  <=50K  
4   Philippines   >50K

 It seems that the '77516' column doesn't synthesize well.

### independent attribute mode

In [15]:
# specify categorical attributes
categorical_attributes_adult = {'United-States': True}

# specify which attributes are candidate keys of input dataset.
candidate_keys_adult = {'39': False}
#unknown

In [16]:
describer_adult_2= DataDescriber(category_threshold=threshold_value_adult)
describer_adult_2.describe_dataset_in_independent_attribute_mode(dataset_file=input_adult,
                                                         attribute_to_is_categorical=categorical_attributes_adult,
                                                         attribute_to_is_candidate_key=candidate_keys_adult)
describer_adult_2.save_dataset_description_to_file('des_adult_2')

In [17]:
generator_adult_2 = DataGenerator()
generator_adult_2.generate_dataset_in_independent_mode(200,'des_adult_2')
generator_adult_2.save_synthetic_data('gen_adult_2')
pd.read_csv('gen_adult_2').head()

39 State-gov     77516  Bachelors  13       Never-married  \
0  42.0   Private  192314.0  Assoc-voc   4       Never-married   
1  49.0   Private  325854.0  Bachelors  10            Divorced   
2  40.0   Private   23894.0    HS-grad   9  Married-civ-spouse   
3  40.0   Private  419230.0       11th   9       Never-married   
4  33.0         ?  342642.0    HS-grad  13            Divorced   

      Adm-clerical  Not-in-family  White    Male    2174      0    40  \
0  Exec-managerial        Husband  White    Male  3990.0   26.0  57.0   
1     Tech-support        Husband  White  Female  3285.0  192.0  54.0   
2            Sales        Husband  White    Male     1.0    9.0  16.0   
3            Sales  Not-in-family  White  Female   910.0   56.0  38.0   
4    Other-service           Wife  White  Female  2534.0  115.0  36.0   

   United-States  <=50K  
0           Iran  <=50K  
1  United-States  <=50K  
2  United-States  <=50K  
3  United-States  <=50K  
4         France   >50K

### correlated attribute mode

In [18]:
# A parameter in Differential Privacy. It roughly means that removing a row in the input dataset will not 
# change the probability of getting the same output more than a multiplicative difference of exp(epsilon).
# Increase epsilon value to reduce the injected noises. Set epsilon=0 to turn off differential privacy.
epsilon = 1

# The maximum number of parents in Bayesian network, i.e., the maximum number of incoming edges. 
degree_of_bayesian_network_adult = 4
#unknown

In [ ]:
describer_adult_3 = DataDescriber(category_threshold=threshold_value_adult)
describer_adult_3.describe_dataset_in_correlated_attribute_mode(dataset_file=input_adult, 
                                                        epsilon=epsilon, 
                                                        k=degree_of_bayesian_network_adult,
                                                        attribute_to_is_categorical=categorical_attributes_adult,
                                                        attribute_to_is_candidate_key=candidate_keys_adult)
describer_adult_3.save_dataset_description_to_file('des_adult_3')

================ Constructing Bayesian Network (BN) ================
Adding ROOT United-States
Adding attribute White
Adding attribute Not-in-family
Adding attribute Never-married
Adding attribute Male
Adding attribute 39
Adding attribute Adm-clerical
Adding attribute Bachelors


In [ ]:
display_bayesian_network(describer_adult_3.bayesian_network)

In [ ]:
generator_adult_3 = DataGenerator()
generator_adult_3.generate_dataset_in_correlated_attribute_mode(200, 'des_adult_3')
generator_adult_3.save_synthetic_data('gen_adult_3')
pd.read_csv('gen_adult_3').head()

# Banknote Authentication data

## SDV

### GaussianCopula Model

In [ ]:
banknote_1= GaussianCopula()
banknote_1.fit(banknote_primary)
banknote_sdv_1=banknote_1.sample(200)
banknote_sdv_1.head()

### CTGAN Model

In [ ]:
banknote_2= CTGAN()
banknote_2.fit(banknote_primary)
banknote_sdv_2=banknote_2.sample(200)
banknote_sdv_2.head()

### CopulaGAN Model

In [ ]:
banknote_3=CopulaGAN()
banknote_3.fit(banknote_primary)
banknote_sdv_3=banknote_3.sample(200)
banknote_sdv_3.head()

### TVAE Model

In [ ]:
banknote_4=TVAE()
banknote_4.fit(banknote_primary)
banknote_sdv_4=banknote_4.sample(200)
banknote_sdv_4.head()

## DataSynthesizer

In [ ]:
input_banknote = 'data_banknote_authentication.txt'

# An attribute is categorical if its domain size is less than this threshold.
threshold_value_banknote= 0

### random mode

In [ ]:
describer_banknote_1= DataDescriber(category_threshold=threshold_value_banknote)
describer_banknote_1.describe_dataset_in_random_mode(input_banknote)
describer_banknote_1.save_dataset_description_to_file('des_banknote_1')

In [ ]:
generator_banknote_1 = DataGenerator()
generator_banknote_1.generate_dataset_in_random_mode(200, 'des_banknote_1')
generator_banknote_1.save_synthetic_data('gen_banknote_1')
pd.read_csv('gen_banknote_1').head()

### independent attribute mode

In [ ]:
# specify categorical attributes
categorical_attributes_banknote = {}

# specify which attributes are candidate keys of input dataset.
candidate_keys_banknote = {'3.6216': False}
#unknown

In [ ]:
describer_banknote_2= DataDescriber(category_threshold=threshold_value_banknote)
describer_banknote_2.describe_dataset_in_independent_attribute_mode(dataset_file=input_banknote,
                                                         attribute_to_is_categorical=categorical_attributes_banknote,
                                                         attribute_to_is_candidate_key=candidate_keys_banknote)
describer_banknote_2.save_dataset_description_to_file('des_banknote_2')

In [ ]:
generator_banknote_2 = DataGenerator()
generator_banknote_2.generate_dataset_in_independent_mode(200,'des_banknote_2')
generator_banknote_2.save_synthetic_data('gen_banknote_2')
pd.read_csv('gen_banknote_2').head()

### correlated attribute mode

In [ ]:
# A parameter in Differential Privacy. It roughly means that removing a row in the input dataset will not 
# change the probability of getting the same output more than a multiplicative difference of exp(epsilon).
# Increase epsilon value to reduce the injected noises. Set epsilon=0 to turn off differential privacy.
epsilon = 1

# The maximum number of parents in Bayesian network, i.e., the maximum number of incoming edges. 
degree_of_bayesian_network_banknote = 2
#unknown

In [ ]:
describer_banknote_3 = DataDescriber(category_threshold=threshold_value_banknote)
describer_banknote_3.describe_dataset_in_correlated_attribute_mode(dataset_file=input_banknote, 
                                                        epsilon=epsilon, 
                                                        k=degree_of_bayesian_network_banknote,
                                                        attribute_to_is_categorical=categorical_attributes_banknote,
                                                        attribute_to_is_candidate_key=candidate_keys_banknote)
describer_banknote_3.save_dataset_description_to_file('des_banknote_3')

In [ ]:
display_bayesian_network(describer_banknote_3.bayesian_network)

In [ ]:
generator_banknote_3 = DataGenerator()
generator_banknote_3.generate_dataset_in_correlated_attribute_mode(200, 'des_banknote_3')
generator_banknote_3.save_synthetic_data('gen_banknote_3')
pd.read_csv('gen_banknote_3').head()

# Titanic data


## SDV

### GaussianCopula Model

In [ ]:
titanic_1= GaussianCopula()
titanic_1.fit(titanic_primary)
titanic_sdv_1=titanic_1.sample(200)
titanic_sdv_1.head()

### CTGAN Model

In [ ]:
titanic_2= CTGAN()
titanic_2.fit(titanic_primary)
titanic_sdv_2=titanic_2.sample(200)
titanic_sdv_2.head()

In [ ]:
### CopulaGAN Model

### CopulaGAN Model

In [ ]:
titanic_3=CopulaGAN()
titanic_3.fit(titanic_primary)
titanic_sdv_3=titanic_3.sample(200)
titanic_sdv_3.head()

### TVAE Model

In [ ]:
titanic_4=TVAE()
titanic_4.fit(titanic_primary)
titanic_sdv_4=titanic_4.sample(200)
titanic_sdv_4.head()

## DataSynthesizer

In [ ]:
input_titanic= 'train_and_test2.csv'

# An attribute is categorical if its domain size is less than this threshold.
threshold_value_titanic= 0

### random mode

In [ ]:
describer_titanic_1= DataDescriber(category_threshold=threshold_value_titanic)
describer_titanic_1.describe_dataset_in_random_mode(input_titanic)
describer_titanic_1.save_dataset_description_to_file('des_titanic_1')

In [ ]:
generator_titanic_1 = DataGenerator()
generator_titanic_1.generate_dataset_in_random_mode(200, 'des_titanic_1')
generator_titanic_1.save_synthetic_data('gen_titanic_1')
pd.read_csv('gen_titanic_1').head()

### independent attribute mode

In [ ]:
# specify categorical attributes
categorical_attributes_titanic = {}

# specify which attributes are candidate keys of input dataset.
candidate_keys_titanic = {'Passengerid': False}
#unknown

In [ ]:
describer_titanic_2= DataDescriber(category_threshold=threshold_value_titanic)
describer_titanic_2.describe_dataset_in_independent_attribute_mode(dataset_file=input_titanic,
                                                         attribute_to_is_categorical=categorical_attributes_titanic,
                                                         attribute_to_is_candidate_key=candidate_keys_titanic)
describer_titanic_2.save_dataset_description_to_file('des_titanic_2')

In [ ]:
generator_titanic_2 = DataGenerator()
generator_titanic_2.generate_dataset_in_independent_mode(200,'des_titanic_2')
generator_titanic_2.save_synthetic_data('gen_titanic_2')
pd.read_csv('gen_titanic_2').head()

### correlated attribute mode

In [ ]:
# A parameter in Differential Privacy. It roughly means that removing a row in the input dataset will not 
# change the probability of getting the same output more than a multiplicative difference of exp(epsilon).
# Increase epsilon value to reduce the injected noises. Set epsilon=0 to turn off differential privacy.
epsilon = 1

# The maximum number of parents in Bayesian network, i.e., the maximum number of incoming edges. 
degree_of_bayesian_network_titanic = 2
#unknown

In [ ]:
describer_titanic_3 = DataDescriber(category_threshold=threshold_value_banknote)
describer_titanic_3.describe_dataset_in_correlated_attribute_mode(dataset_file=input_titanic, 
                                                        epsilon=epsilon, 
                                                        k=degree_of_bayesian_network_titanic,
                                                        attribute_to_is_categorical=categorical_attributes_titanic,
                                                        attribute_to_is_candidate_key=candidate_keys_titanic)
describer_titanic_3.save_dataset_description_to_file('des_titanic_3')

In [ ]:
display_bayesian_network(describer_titanic_3.bayesian_network)

In [ ]:
generator_titanic_3 = DataGenerator()
generator_titanic_3.generate_dataset_in_correlated_attribute_mode(200, 'des_titanic_3')
generator_titanic_3.save_synthetic_data('gen_titanic_3')
pd.read_csv('gen_titanic_3').head()